libsolidity/experimental/ast/TypeSystem.h

```diff
@@ -0,0 +1,184 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <libsolidity/experimental/ast/Type.h>
+#include <liblangutil/Exceptions.h>
+
+#include <optional>
+#include <string>
+#include <variant>
+#include <vector>
+
+namespace solidity::frontend
+{
+class Declaration;
+}
+
+namespace solidity::frontend::experimental
+{
+
+class TypeEnvironment
+{
+public:
+	struct TypeMismatch
+	{
+		Type a;
+		Type b;
+	};
+
+	struct SortMismatch {
+		Type type;
+		Sort sort;
+	};
+
+	struct RecursiveUnification
+	{
+		Type var;
+		Type type;
+	};
+
+	using UnificationFailure = std::variant<
+		TypeMismatch,
+		SortMismatch,
+		RecursiveUnification
+	>;
+
+	TypeEnvironment(TypeSystem& _typeSystem): m_typeSystem(_typeSystem) {}
+	TypeEnvironment(TypeEnvironment const&) = delete;
+	TypeEnvironment& operator=(TypeEnvironment const&) = delete;
+	TypeEnvironment clone() const;
+
+	Type resolve(Type _type) const;
+	Type resolveRecursive(Type _type) const;
+	Type fresh(Type _type);
+	[[nodiscard]] std::vector<UnificationFailure> unify(Type _a, Type _b);
+	Sort sort(Type _type) const;
+	bool typeEquals(Type _lhs, Type _rhs) const;
+
+	TypeSystem& typeSystem() { return m_typeSystem; }
+	TypeSystem const& typeSystem() const { return m_typeSystem; }
+
+private:
+	TypeEnvironment(TypeEnvironment&& _env):
+		m_typeSystem(_env.m_typeSystem),
+		m_typeVariables(std::move(_env.m_typeVariables))
+	{}
+
+	[[nodiscard]] std::vector<TypeEnvironment::UnificationFailure> instantiate(TypeVariable _variable, Type _type);
+
+	TypeSystem& m_typeSystem;
+	std::map<size_t, Type> m_typeVariables;
+};
+
+class TypeSystem
+{
+public:
+	struct TypeConstructorInfo
+	{
+		std::string name;
+		std::string canonicalName;
+		std::vector<Arity> arities;
+		Declaration const* typeDeclaration = nullptr;
+		size_t arguments() const
+		{
+			solAssert(!arities.empty());
+			return arities.front().argumentSorts.size();
+		}
```

Comment:
Is the first element in `arities` meant to be always `PrimitiveClass::Type`? Can we assert that? If not, what would a different type class on the first position mean?

Can a type constructor have more than one `PrimitiveClass::Type` in `arities`? If so, what would that mean?

If there's always exactly one, then it seems to me that it we should have a separate field for it and not just lump it with other things in `arities`. It seems different in some ways and it's even excluded by some checks.

Reply:
Oh, and by the way calling this thing `Arity` confused me a lot. Looking at other code I assumed it was just something describing the number of args, but it looks like an equivalent of `TypeConstant` but for type classes. I'm only now noticing that it has a type class inside, which now exlains `TypeSystem::instantiateClass()` (it was a bit weird that you'd not need to say which class you're instantiating).

Reply:
It's how these things are called :-). If you can come up with a better name, feel free to suggest it.
It describes for a type constructor that applying it to type arguments of specific sorts (the argument sorts in the arity), the result will have a specific class (the class in the arity).